mqtt/mqtt-bridge/src/persist/waking_state/waking_store.rs

```diff
@@ -0,0 +1,97 @@
+use std::collections::hash_map::Entry;
+use std::{boxed::Box, collections::HashMap, task::Waker};
+
+use bincode::{deserialize, serialize, ErrorKind};
+use mqtt3::proto::Publication;
+use rocksdb::{IteratorMode, DB};
+use tracing::error;
+
+use crate::persist::{waking_state::StreamWakeableState, Key, PersistError};
+
+/// Responsible for waking waiting streams when new elements are added.
+///
+/// Exposes a get method for retrieving a count of elements in order of insertion.
+/// When elements are retrieved they are added to the in flight collection, but kept in the original db store.
+/// Only when elements are removed from the in-flight collection they will be removed from the store.
+pub struct WakingStore {
+    db: DB,
+    in_flight: HashMap<Key, Publication>,
+    waker: Option<Waker>,
+}
+
+impl WakingStore {
+    pub fn new(db: DB) -> Self {
+        let in_flight = HashMap::new();
+        let waker = None;
+
+        Self {
+            db,
+            in_flight,
+            waker,
+        }
+    }
+}
+
+impl StreamWakeableState for WakingStore {
+    fn insert(&mut self, key: Key, value: Publication) -> Result<(), PersistError> {
+        let key_bytes = serialize(&key).map_err(PersistError::Serialization)?;
+        let publication_bytes = serialize(&value).map_err(PersistError::Serialization)?;
+        self.db
+            .put(key_bytes, publication_bytes)
+            .map_err(PersistError::Insertion)?;
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+
+    /// Get count elements of store, exluding those that are already in in-flight
+    fn get(&mut self, count: usize) -> Vec<(Key, Publication)> {
```

Comment:
I think it should return Result and propagade error up the callstack

Reply:
something lile this
```rust
fn get(&mut self, count: usize) -> Result<Vec<(Key, Publication)>, PersistError> {
        let iter = self.db.iterator(IteratorMode::Start);

        iter.map(|(key, publication)| {
            bincode::deserialize(&key)
                .map_err(PersistError::Serialization)
                .and_then(|key| {
                    bincode::deserialize(&publication)
                        .map_err(PersistError::Serialization)
                        .map(|publication| (key, publication))
                })
        })
        .collect()
    }
        ```

Reply:
If we go with this approach then we can't deserialize anything if there is a failure. I think that you are correct though in that if this is a logic error it shouldn't matter. We should percolate up and blow up.

Reply:
The issue though is that this batch function is used in the loader. The loader implements stream, and this batch() func is used in the poll_next() impl for stream. 

Since poll_next doesn't return result, it cannot get percolated all the way up. Should we still return a Result here and log an error in the stream implementation? I don't see an easy way we can blow up without a bigger change to the structure.

Reply:
good question. if we have an error on this step it is either someone forged the database or we the code didn't expect database schema change. in this case, we should fail fast and loud

Reply:
Ok that makes sense. Should we panic? Also, I remember there was a macro for failing with a message? I looked around and can't find it.

Reply:
do you ask about https://doc.rust-lang.org/std/result/enum.Result.html#method.expect ?

Reply:
No I remember seeing something different. Is it ok to use expect here?

Reply:
I am using expect